src/renderer/RenderLayer.js

```diff
@@ -60,17 +63,90 @@ export default class RenderLayer {
         if (!this.viz) {
             return [];
         }
-        return [].concat(...this.getActiveDataframes().map(df =>
+
+        const noPossiblePartialFeatures = (this.viz.geometryType === GEOMETRY_TYPE.POINT);
+        if (noPossiblePartialFeatures) {
+            return this._getPointFeaturesAtPosition(pos);
+        } else {
+            return this._getPartialFeaturesAtPosition(pos);
+        }
+    }
+
+    _getPointFeaturesAtPosition (pos) {
+        const rawFeatures = this._getRawFeaturesAtPosition(pos);
+        return rawFeatures.map((raw) => { return this._buildFeatureFromRaw(raw); });
+    }
+
+    _buildFeatureFromRaw (rawFeature) {
+        const { viz, customizedFeatures, trackFeatureViz, idProperty } = this;
+        const featureVizParams = { viz, customizedFeatures, trackFeatureViz, idProperty };
+        return new Feature(rawFeature, featureVizParams);
+    }
+
+    _getRawFeaturesAtPosition (pos) {
+        const rawFeatures = [].concat(...this.getActiveDataframes().map(df =>
             df.getFeaturesAtPosition(pos, this.viz)
-        )).map(rawFeature => {
-            const featureVizParams = {
-                viz: this.viz,
-                customizedFeatures: this.customizedFeatures,
-                trackFeatureViz: this.trackFeatureViz,
-                idProperty: this.idProperty
-            };
-            return new Feature(rawFeature, featureVizParams);
+        ));
+        return rawFeatures;
+    }
+
+    _getPartialFeaturesAtPosition (pos) {
+        const rawFeatures = this._getRawFeaturesAtPosition(pos);
+        if (rawFeatures.length === 0) return [];
+
+        const rawPartialFeatures = this._getPartialFeaturesFromSingle(rawFeatures);
+
+        const compoundFeatures = this._getCompoundFeaturesFrom(rawPartialFeatures);
+        return compoundFeatures;
+    }
+
+    _getCompoundFeaturesFrom (rawPartialFeatures) {
+        const features = [];
+        for (let featureId in rawPartialFeatures) {
+            const viewporFeaturePieces = rawPartialFeatures[featureId];
+            const featurePieces = viewporFeaturePieces.map((raw) => { return this._buildFeatureFromRaw(raw); });
+            features.push(getCompoundFeature(featurePieces));
+        }
+        return features;
+    }
+
+    /**
+     * Get all the pieces from rawFeatures.
+     * Return is a nested Array; each entry includes an array with (potentially) several feature pieces
+     */
+    _getPartialFeaturesFromSingle (rawFeatures) {
+        const idProperty = this.viz.metadata.idProperty;
```

Comment:
Should we use `this.idProperty`? Or is it unsafe?

Reply:
Good point, let me check